### PR TITLE
HCX-97 - Handle error reported by the recipient post decryption

### PR DESCRIPTION
**Pull request by tarento-dharmateja**

### CODE REVIEW

File: api-gateway/src/main/java/org/swasth/apigateway/filters/HCXValidationFilter.java [hunk 1]
```diff
@@ -58,10 +60,16 @@ public GatewayFilter apply(Config config) {
...
+                Map<String,Object> requestBody = JSONUtils.deserialize(cachedBody.toString(), HashMap.class);
+                if(requestBody.containsKey(Constants.PAYLOAD)) {
+                    Request request = new Request(requestBody);
+                    correlationId = request.getCorrelationId();
+                    apiCallId = request.getApiCallId();
+                    request.validate(getMandatoryHeaders(), getDetails(request.getSenderCode()), getDetails(request.getRecipientCode()), getSubject(exchange), timestampRange);
+                } else {
+                    ErrorRequest request = new ErrorRequest(requestBody);
+                    request.validate(getErrorMandatoryHeaders(), getDetails((String) requestBody.get(Constants.SENDER_CODE)),getDetails((String) requestBody.get(Constants.RECIPIENT_CODE)), getSubject(exchange));
+                }
```
Comment (maheshkumargangula) [inline api-gateway/src/main/java/org/swasth/apigateway/filters/HCXValidationFilter.java:72]: Rename the "Request" class as "JWERequest". 
Rename the "ErrorRequest" class as "PlainRequest".

PlainRequst class should have implementation of getSenderCode, getRecipientCode methods.

File: api-gateway/src/main/java/org/swasth/apigateway/models/ErrorRequest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    public void validate(List<String> mandatoryHeaders,Map<String, Object> senderDetails, Map<String, Object> recipientDetails, String subject) throws ClientException {
+        List<String> missingHeaders = mandatoryHeaders.stream().filter(key -> !errorRequest.containsKey(key)).collect(Collectors.toList());
+        if (!missingHeaders.isEmpty()) {
+            throw new ClientException(ErrorCodes.ERR_MANDATORY_HEADERFIELD_MISSING, "Mandatory headers are missing: " + missingHeaders);
+        }
+
+
+        String correlationId = (String) errorRequest.get(Constants.CORRELATION_ID);
```
Comment (maheshkumargangula) [inline api-gateway/src/main/java/org/swasth/apigateway/models/ErrorRequest.java:37]: Implement getCorrelationId as a method in pojo.

File: api-gateway/src/main/java/org/swasth/apigateway/models/ErrorRequest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        validateCondition(!Utils.isUUID(correlationId), ErrorCodes.ERR_INVALID_CORRELATION_ID, "Correlation id should be a valid UUID");
+
+        String status = (String) errorRequest.get(Constants.STATUS);
```
Comment (maheshkumargangula) [inline api-gateway/src/main/java/org/swasth/apigateway/models/ErrorRequest.java:40]: Same as CorrelationId

File: hcx-apis/src/main/java/org/swasth/hcx/helpers/EventGenerator.java [hunk 4]
```diff
@@ -27,34 +27,53 @@ public String generatePayloadEvent(String mid, Request request) throws JsonProce
...
+            event.put(ACTION, apiAction);
+            event.put(HEADERS, new HashMap<>(){{
+                put(JOSE, filterJoseHeaders);
+                put(PROTOCOL, filterProtocolHeaders);
+            }});
+            event.put(LOG_DETAILS, new HashMap<>(){{
+                put(CODE, "");
+                put(MESSAGE, "");
+                put(TRACE, "");
+            }});
```
Comment (maheshkumargangula) [inline hcx-apis/src/main/java/org/swasth/hcx/helpers/EventGenerator.java:58]: Don't put empty data.

File: hcx-core/hcx-common/src/main/java/org/swasth/common/utils/Constants.java [hunk 5]
```diff
@@ -53,6 +53,8 @@
 
     //Request props
     public static final String PROTOCOL_HEADERS_MANDATORY = "protocol.headers.mandatory";
+    public static final String ERROR_HEADERS_MANDATORY = "error.headers.mandatory";
+    public static final String ERROR_HEADERS_OPTIONAL = "error.headers.optional";
```
Comment (maheshkumargangula) [inline hcx-core/hcx-common/src/main/java/org/swasth/common/utils/Constants.java:57]: Change the name. Maybe "plainrequest.headers.mandatory"

**Review by maheshkumargangula — CHANGES_REQUESTED** [review]
Improve the code coverage.

File: api-gateway/src/main/java/org/swasth/apigateway/filters/HCXValidationFilter.java (hunk 1 above)
Comment (tarento-dharmateja) [reply api-gateway/src/main/java/org/swasth/apigateway/filters/HCXValidationFilter.java]: Done

File: api-gateway/src/main/java/org/swasth/apigateway/models/ErrorRequest.java (hunk 2 above)
Comment (tarento-dharmateja) [reply api-gateway/src/main/java/org/swasth/apigateway/models/ErrorRequest.java]: Done

File: hcx-apis/src/main/java/org/swasth/hcx/helpers/EventGenerator.java (hunk 4 above)
Comment (tarento-dharmateja) [reply hcx-apis/src/main/java/org/swasth/hcx/helpers/EventGenerator.java]: Done

File: api-gateway/src/main/java/org/swasth/apigateway/models/ErrorRequest.java (hunk 3 above)
Comment (tarento-dharmateja) [reply api-gateway/src/main/java/org/swasth/apigateway/models/ErrorRequest.java]: Done

File: hcx-core/hcx-common/src/main/java/org/swasth/common/utils/Constants.java (hunk 5 above)
Comment (tarento-dharmateja) [reply hcx-core/hcx-common/src/main/java/org/swasth/common/utils/Constants.java]: Done

**Comment by maheshkumargangula** [comment]
There is still scope to improve the code coverage.

**Comment by tarento-dharmateja** [comment]
> There is still scope to improve the code coverage.

We have not yet written test cases for the hcx core service, I checked sonar cloud for missing coverage hcx-core is missing.

**Pull request merged**